### PR TITLE
Move remaining Matrix tests to matrixlib

### PR DESCRIPTION
**Pull request by mhvk**
Continuing #10975 and #10986, this moves all remaining tests that involve `np.matrix` to `matrixlib` itself. The only dependencies left are in explicit checks in the code.

### CODE REVIEW

File: numpy/core/tests/test_nditer.py [hunk 1]
```diff
@@ -1469,26 +1469,25 @@ def test_iter_allocate_output_types_scalar():
...
-    # but if subtypes are disabled, the result can still work
+    # If subtypes are disabled, we should get back an ndarray.
     i = nditer([a, b, None], [],
-            [['readonly'], ['readonly'], ['writeonly', 'allocate', 'no_subtype']])
+               [['readonly'], ['readonly'],
+                ['writeonly', 'allocate', 'no_subtype']])
     assert_equal(type(b), type(i.operands[2]))
-    assert_(type(a) != type(i.operands[2]))
-    assert_equal(i.operands[2].shape, (1, 2, 2))
+    assert_(type(a) is not (i.operands[2]))
```
Comment (eric-wieser) [inline numpy/core/tests/test_nditer.py]: Missing `type` here

Comment (mhvk) [reply numpy/core/tests/test_nditer.py]: Oops. And then the test of course still passes...

Comment (charris) [reply numpy/core/tests/test_nditer.py]: Still needs fixing.

**Comment by mhvk** [comment]
Strange, I could have sworn I had made that change already! But it looks like I forgot to push it up...  Anyway, tests now pass, so all looks OK.

File: numpy/core/fromnumeric.py [hunk 2]
```diff
@@ -1336,10 +1336,10 @@ def diagonal(a, offset=0, axis1=0, axis2=1):
     Returns
     -------
     array_of_diagonals : ndarray
-        If `a` is 2-D and not a `matrix`, a 1-D array of the same type as `a`
-        containing the diagonal is returned. If `a` is a `matrix`, a 1-D
+        If `a` is 2-D, a 1-D array of the same type as `a` containing the
+        diagonal is returned (except if `a` is a `matrix`, in which case a 1-D
```
Comment (charris) [inline numpy/core/fromnumeric.py]: "is returned unless `a` is a matrix, in which ..."

File: numpy/core/fromnumeric.py [hunk 3]
```diff
@@ -1336,10 +1336,10 @@ def diagonal(a, offset=0, axis1=0, axis2=1):
...
         array containing the diagonal is returned in order to maintain
-        backward compatibility.
+        backward compatibility).
```
Comment (charris) [inline numpy/core/fromnumeric.py]: And no closing `)`.

File: numpy/core/fromnumeric.py [hunk 4]
```diff
@@ -1496,10 +1496,9 @@ def ravel(a, order='C'):
     Returns
     -------
     y : array_like
-        If `a` is a matrix, y is a 1-D ndarray, otherwise y is an array of
-        the same subtype as `a`. The shape of the returned array is
-        ``(a.size,)``. Matrices are special cased for backward
-        compatibility.
+        y is an array of the same subtype as `a`, with shape ``(a.size,)``
+        (Note: matrices are special-cases for backward compatibility: if `a`
```
Comment (charris) [inline numpy/core/fromnumeric.py]: Don't parenthesize. "Note that matrices ..., if `a` ..." .

File: numpy/matrixlib/tests/test_interaction.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    # Test to make sure matrices give the same answer as ndarrays
+    # 2018-04-29: moved here from core.tests.test_function_base.
+    x = np.linspace(0, 5)
+    y = x * x
+    r = np.trapz(y, x)
+    mx = np.matrix(x)
+    my = np.matrix(y)
+    mr = np.trapz(my, mx)
+    assert_almost_equal(mr, r)
+
```
Comment (charris) [inline numpy/matrixlib/tests/test_interaction.py]: Should we check the type of `mr` also?

File: numpy/matrixlib/tests/test_interaction.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    result = np.apply_along_axis(double, 1, m)
+    assert_(isinstance(result, np.matrix))
+    assert_array_equal(result, expected)
+
+
+def test_kron_matrix():
+    # 2018-04-29: moved here from core.tests.test_shape_base.
+    a = np.ones([2, 2])
+    m = np.asmatrix(a)
+    assert_equal(type(np.kron(a, a)), np.ndarray)
```
Comment (charris) [inline numpy/matrixlib/tests/test_interaction.py]: You used `is` above for checking types, want to leave this as is?

File: numpy/matrixlib/tests/test_interaction.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        assert_equal(type(ab_c), np.matrix)
+
+        assert_equal(np.array(ab_r), [[1, 2, 3, 4]])
+        assert_equal(np.array(ab_c), [[1], [2], [3], [4]])
+
+        assert_raises(ValueError, lambda: np.r_['rc', a, b])
+
+    def test_matrix_scalar(self):
+        r = np.r_['r', [1, 2], 3]
+        assert_equal(type(r), np.matrix)
```
Comment (charris) [inline numpy/matrixlib/tests/test_interaction.py]: See above. I really don't know if there is a preference between the two ways to check equality of types.

**Comment by charris** [comment]
I fixed the conflicts, so you may want to pull this before doing more work. I'll change the docstrings if you wish.

**Comment by mhvk** [comment]
@charris - Thanks for the comments. Happy to make the changes, though as I've got to write some reference letters now, it will be tomorrow (so fine if you want to just do them).

As for `is` vs `==`, I think in principle the recommendation is to test class identity using `is`, but on the other hand it is good to not change more than necessary. I'm afraid I have not been very consistent in finding a balance between the two...

File: numpy/core/fromnumeric.py [hunk 8]
```diff
@@ -1336,10 +1336,11 @@ def diagonal(a, offset=0, axis1=0, axis2=1):
     Returns
     -------
     array_of_diagonals : ndarray
-        If `a` is 2-D and not a `matrix`, a 1-D array of the same type as `a`
-        containing the diagonal is returned. If `a` is a `matrix`, a 1-D
-        array containing the diagonal is returned in order to maintain
-        backward compatibility.
+        If `a` is 2-D, then a 1-D array containing the diagonal and of the
+        same type as `a` is returned unless `a` is a `matrix`, in which case
```
Comment (charris) [inline numpy/core/fromnumeric.py]: `type`, `subtype`, or both?

**Comment by charris** [comment]
OK, made the changes but will wait on feedback before putting this in.

File: numpy/core/fromnumeric.py (hunk 8 above)
Comment (mhvk) [reply numpy/core/fromnumeric.py]: the test is with `isinstance`

**Comment by charris** [comment]
Thanks Marten.

**Pull request merged**